Shipyard-controller 0.14.2 crashes with panic

<!--
For Security vulnerabilities, please refer to our security page https://github.com/keptn/keptn/blob/master/SECURITY.md
-->

**Environment**

* Client OS (e.g., Linux, macOS, Windows): Ubuntu-20.04 (GitHub runner)
* Keptn Version (`keptn version`): 0.14.2
* Kubernetes Cloud Provider (e.g., GKE, AKS): v1.21.12+k3s1
* Kubernetes version (`kubectl version`): -

<!-- Please also let us know about other components and their version if you believe that this issue is related to them (e.g., Istio, Web Browser Version). -->

**Affected Component**
<!-- Please tell us which component of Keptn is affected (leave empty if you are unsure). -->

* [ ] Approval
* [ ] Bridge
* [ ] CLI
* [ ] Datastore
* [ ] Docs
* [ ] Distributor / Event Broker
* [ ] Helm
* [ ] Istio
* [ ] Jmeter
* [ ] Kubernetes Integration
* [ ] Openshift Integration
* [ ] REST API
* [ ] Webhook

**Describe the bug**
Sometimes the integration tests for the job-executor-service are not successful, because nginx returns a 502 Bad Gateway error as shown in this [run](https://github.com/keptn-contrib/job-executor-service/runs/7105245137?check_suite_focus=true#step:13:57).

This seem to happen when deleting a project:
```
2022/06/29 04:10:25 [error] 33#33: *921 upstream prematurely closed connection while reading response header from upstream, client: 10.42.0.1, server: _, request: "DELETE /api/controlPlane/v1/project/jes-locust HTTP/1.1", upstream: "http://10.43.218.236:8080/v1/project/jes-locust", host: "10.1.0.93"
10.42.0.1 - - [29/Jun/2022:04:10:25 +0000] "DELETE /api/controlPlane/v1/project/jes-locust HTTP/1.1" 502 150 "-" "Go-http-client/1.1"
2022/06/29 04:10:25 [error] 35#35: *1054 connect() failed (111: Connection refused) while connecting to upstream, client: 10.42.0.1, server: _, request: "POST /api/controlPlane/v1/project HTTP/1.1", upstream: "http://10.43.218.236:8080/v1/project", host: "10.1.0.93"
```

The cause of that behavior seems to be a crash of the shipyard-controller:
```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x48 pc=0x162c9ec]

goroutine 5822 [running]:
github.com/keptn/keptn/shipyard-controller/handler.(*ShipyardRetriever).GetCachedShipyard(0xeda4dc7b1, {0xc00048a850, 0x0})
	/go/src/github.com/keptn/keptn/shipyard-controller/handler/shipyardretriever.go:71 +0x2c
github.com/keptn/keptn/shipyard-controller/handler.(*shipyardController).triggerNextTaskSequences(_, {{{0xc00048a850, 0xa}, {0xc00048a86c, 0x2}, {0xc00048a860, 0xc}, 0x0, {0x0, 0x0}, ...}, ...}, ...)
	/go/src/github.com/keptn/keptn/shipyard-controller/handler/shipyardcontroller.go:720 +0x63
github.com/keptn/keptn/shipyard-controller/handler.(*shipyardController).proceedTaskSequence(_, {{{0xc00048a850, 0xa}, {0xc00048a86c, 0x2}, {0xc00048a860, 0xc}, 0x0, {0x0, 0x0}, ...}, ...}, ...)
	/go/src/github.com/keptn/keptn/shipyard-controller/handler/shipyardcontroller.go:699 +0x308
github.com/keptn/keptn/shipyard-controller/handler.(*shipyardController).onTaskProgress(_, {{0x0, 0x0}, {0x18927c0, 0xc000803f80}, {0x0, 0x0}, {0xc0006af920, 0x24}, {0xc0006af950, ...}, ...}, ...)
	/go/src/github.com/keptn/keptn/shipyard-controller/handler/shipyardcontroller.go:373 +0x5e5
github.com/keptn/keptn/shipyard-controller/handler.(*shipyardController).handleTaskEvent(0xc000103860, {{0x0, 0x0}, {0x18927c0, 0xc000803f80}, {0x0, 0x0}, {0xc0006af920, 0x24}, {0xc0006af950, ...}, ...})
	/go/src/github.com/keptn/keptn/shipyard-controller/handler/shipyardcontroller.go:316 +0x2e8
github.com/keptn/keptn/shipyard-controller/handler.(*shipyardController).HandleIncomingEvent.func2()
	/go/src/github.com/keptn/keptn/shipyard-controller/handler/shipyardcontroller.go:168 +0x65
created by github.com/keptn/keptn/shipyard-controller/handler.(*shipyardController).HandleIncomingEvent
	/go/src/github.com/keptn/keptn/shipyard-controller/handler/shipyardcontroller.go:167 +0x4d1

[...]
```

For the full context please see the have a look at [keptn-support-archive-0.14.2-netpolicy-false-jobnetpolicy-true.zip](https://github.com/keptn/keptn/files/9008303/keptn-support-archive-0.14.2-netpolicy-false-jobnetpolicy-true.zip)

**To Reproduce**
<!-- Please provide detailed instructions on how to reproduce the behaviour, including code samples if applicable. -->
I'm not able to consistently reproduce the error locally or with any other integration test runs

**Expected behavior**
<!--- Please provide a clear and concise description of what you expected to happen -->
Project is deleted and the shipyard-controller can still process other requests without crashing

**Current behavior**
<!--- Please tell us what happens instead of the expected behavior -->
The shipyard-controller crashes and renders the keptn installation unusable for the remaining time of the integration test run

--- a/shipyard-controller/db/mongodb_project_repo.go
+++ b/shipyard-controller/db/mongodb_project_repo.go
@@ -62,8 +62,11 @@ func (mdbrepo *MongoDBProjectsRepo) GetProject(projectName string) (*apimodels.E
 
 	projectCollection := mdbrepo.getProjectsCollection()
 	result := projectCollection.FindOne(ctx, bson.M{"projectName": projectName})
-	if result.Err() != nil && result.Err() == mongo.ErrNoDocuments {
-		return nil, nil
+	if result.Err() != nil {
+		if result.Err() == mongo.ErrNoDocuments {
+			return nil, ErrProjectNotFound
+		}
+		return nil, result.Err()
 	}
 	projectResult := &apimodels.ExpandedProject{}
 	err = result.Decode(projectResult)

--- a/shipyard-controller/db/mongodb_project_repo_test.go
+++ b/shipyard-controller/db/mongodb_project_repo_test.go
@@ -1,0 +1,77 @@
+package db
+
+import (
+	"fmt"
+	apimodels "github.com/keptn/go-utils/pkg/api/models"
+	"github.com/stretchr/testify/require"
+	"testing"
+)
+
+func TestMongoDBProjectsRepo_InsertAndRetrieve(t *testing.T) {
+	r := NewMongoDBProjectsRepo(GetMongoDBConnectionInstance())
+
+	err := r.CreateProject(&apimodels.ExpandedProject{
+		ProjectName: "my-project",
+	})
+
+	require.Nil(t, err)
+
+	prj, err := r.GetProject("my-project")
+
+	require.Nil(t, err)
+	require.NotNil(t, prj)
+
+	err = r.DeleteProject("my-project")
+	require.Nil(t, err)
+
+	prj, err = r.GetProject("my-project")
+
+	require.ErrorIs(t, err, ErrProjectNotFound)
+	require.Nil(t, prj)
+}
+
+func TestMongoDBProjectsRepo_InsertAndRetrieveMultiple(t *testing.T) {
+	r := NewMongoDBProjectsRepo(GetMongoDBConnectionInstance())
+
+	nrProjects := 5
+
+	for i := 0; i < nrProjects; i++ {
+		projectName := fmt.Sprintf("project-%d", i)
+		err := r.CreateProject(&apimodels.ExpandedProject{
+			ProjectName: projectName,
+		})
+
+		require.Nil(t, err)
+	}
+
+	prj, err := r.GetProjects()
+
+	require.Nil(t, err)
+	require.NotNil(t, prj)
+	require.Len(t, prj, 5)
+}
+
+func TestMongoDBProjectsRepo_UpdateProject(t *testing.T) {
+	r := NewMongoDBProjectsRepo(GetMongoDBConnectionInstance())
+
+	err := r.CreateProject(&apimodels.ExpandedProject{
+		ProjectName: "my-project",
+	})
+
+	require.Nil(t, err)
+
+	updatedProject := &apimodels.ExpandedProject{
+		ProjectName: "my-project",
+		Shipyard:    "shipyard-content",
+	}
+	err = r.UpdateProject(updatedProject)
+
+	require.Nil(t, err)
+
+	prj, err := r.GetProject("my-project")
+
+	require.Nil(t, err)
+	require.NotNil(t, prj)
+
+	require.Equal(t, updatedProject, prj)
+}

--- a/shipyard-controller/handler/projectmanager.go
+++ b/shipyard-controller/handler/projectmanager.go
@@ -86,7 +86,7 @@ func (pm *ProjectManager) GetByName(projectName string) (*apimodels.ExpandedProj
 func (pm *ProjectManager) Create(params *models.CreateProjectParams) (error, common.RollbackFunc) {
 
 	if err := pm.checkForExistingProject(params); err != nil {
-		return err, nilRollback
+		return fmt.Errorf("could not create project '%s': %w", *params.Name, err), nilRollback
 	}
 
 	err := pm.updateGITRepositorySecret(*params.Name, decodeGitCredentials(params.GitCredentials))
@@ -165,7 +165,7 @@ func (pm *ProjectManager) checkForExistingProject(params *models.CreateProjectPa
 	existingProject, err := pm.ProjectMaterializedView.GetProject(*params.Name)
 	if err != nil && err != db.ErrProjectNotFound {
 		log.Errorf("Error occurred while getting project: %s", err.Error())
-		return fmt.Errorf("failed to get project: '%s'", *params.Name)
+		return fmt.Errorf("failed to get information for project '%s'", *params.Name)
 	}
 	if existingProject != nil {
 		return ErrProjectAlreadyExists

--- a/shipyard-controller/handler/projectmanager.go
+++ b/shipyard-controller/handler/projectmanager.go
@@ -85,16 +85,11 @@ func (pm *ProjectManager) GetByName(projectName string) (*apimodels.ExpandedProj
 
 func (pm *ProjectManager) Create(params *models.CreateProjectParams) (error, common.RollbackFunc) {
 
-	existingProject, err := pm.ProjectMaterializedView.GetProject(*params.Name)
-	if err != nil && err != db.ErrProjectNotFound {
-		log.Errorf("Error occurred while getting project: %s", err.Error())
-		return fmt.Errorf("failed to get project: '%s'", *params.Name), nilRollback
-	}
-	if existingProject != nil {
-		return ErrProjectAlreadyExists, nilRollback
+	if err := pm.checkForExistingProject(params); err != nil {
+		return err, nilRollback
 	}
 
-	err = pm.updateGITRepositorySecret(*params.Name, decodeGitCredentials(params.GitCredentials))
+	err := pm.updateGITRepositorySecret(*params.Name, decodeGitCredentials(params.GitCredentials))
 	if err != nil {
 		return err, nilRollback
 	}
@@ -164,6 +159,18 @@ func (pm *ProjectManager) Create(params *models.CreateProjectParams) (error, com
 	pm.deleteProjectSequenceCollections(*params.Name)
 
 	return nil, nilRollback
+}
+
+func (pm *ProjectManager) checkForExistingProject(params *models.CreateProjectParams) error {
+	existingProject, err := pm.ProjectMaterializedView.GetProject(*params.Name)
+	if err != nil && err != db.ErrProjectNotFound {
+		log.Errorf("Error occurred while getting project: %s", err.Error())
+		return fmt.Errorf("failed to get project: '%s'", *params.Name)
+	}
+	if existingProject != nil {
+		return ErrProjectAlreadyExists
+	}
+	return nil
 }
 
 func (pm *ProjectManager) Update(params *models.UpdateProjectParams) (error, common.RollbackFunc) {

--- a/shipyard-controller/handler/projectmanager.go
+++ b/shipyard-controller/handler/projectmanager.go
@@ -86,7 +86,7 @@ func (pm *ProjectManager) GetByName(projectName string) (*apimodels.ExpandedProj
 func (pm *ProjectManager) Create(params *models.CreateProjectParams) (error, common.RollbackFunc) {
 
 	existingProject, err := pm.ProjectMaterializedView.GetProject(*params.Name)
-	if err != nil {
+	if err != nil && err != db.ErrProjectNotFound {
 		log.Errorf("Error occurred while getting project: %s", err.Error())
 		return fmt.Errorf("failed to get project: '%s'", *params.Name), nilRollback
 	}

--- a/shipyard-controller/handler/projectmanager.go
+++ b/shipyard-controller/handler/projectmanager.go
@@ -165,7 +165,7 @@ func (pm *ProjectManager) checkForExistingProject(params *models.CreateProjectPa
 	existingProject, err := pm.ProjectMaterializedView.GetProject(*params.Name)
 	if err != nil && err != db.ErrProjectNotFound {
 		log.Errorf("Error occurred while getting project: %s", err.Error())
-		return fmt.Errorf("failed to get information for project '%s'", *params.Name)
+		return fmt.Errorf("failed to get information for project '%s': %w", *params.Name, err)
 	}
 	if existingProject != nil {
 		return ErrProjectAlreadyExists

